smarts/core/argoverse_map.py

```diff
@@ -948,11 +964,42 @@ def waypoint_paths(
             road_ids = [road.road_id for road in route.roads]
         if road_ids:
             return self._waypoint_paths_along_route(pose.point, lookahead, road_ids)
+
         closest_lps = self._lanepoints.closest_lanepoints(
             [pose], within_radius=within_radius
         )
-        closest_lane = closest_lps[0].lane
+        closest_lane: RoadMap.Lane = closest_lps[0].lane
+
         waypoint_paths = []
+        if closest_lane.in_junction:
+            junction_lane = closest_lane
+            closest_lanes = [lp.lane for lp in closest_lps]
+            if self._last_junction_lane and self._last_junction_lane in closest_lanes:
+                junction_lane = self._last_junction_lane
+            paths = self._resolve_in_junction(junction_lane)
+            for road_ids in paths:
+                new_paths = self._waypoint_paths_along_route(
+                    pose.point, lookahead, road_ids
+                )
+                for path in new_paths:
+
+                    def _angle_between(pose, wp):
+                        heading_vec = pose.heading.direction_vector()
+                        wp_vec = wp.pos - pose.position[:2]
+                        angle = np.arccos(
+                            np.dot(heading_vec, wp_vec)
+                            / (np.linalg.norm(heading_vec) * np.linalg.norm(wp_vec))
+                        )
+                        return angle
+
+                    # Filter out any waypoints that are behind the vehicle
+                    wps = [wp for wp in path if _angle_between(pose, wp) < np.pi / 1.5]
+                    if len(wps) > 0:
+                        waypoint_paths.append(wps)
+            self._last_junction_lane = junction_lane
```

Comment:
Is this intended to be primitive caching? Can this handle agents simultaneously in multiple intersections?

Reply:
I've removed this.